rosetta-source/src/main/java/cdm/base/staticdata/codelist/ValidateFpMLCodingSchemeImpl.java

```diff
@@ -0,0 +1,155 @@
+
+        // Load the JSON file as an InputStream from the classpath
+        try (InputStream inputStream = classLoader.getResourceAsStream(matchedFile)) {
+            if (inputStream == null) {
+                throw new FileNotFoundException("Failed to load resource: " + matchedFile);
+            }
+            return mapper.readValue(inputStream, CodeList.class);
+        }
+    }
+
+    /**
+     * Retrieves a set of JSON filenames from the CodeList directory.
+     *
+     * This method detects whether the resources are being accessed from a JAR file or from a standard
+     * file system (development mode). It lists the JSON files accordingly.
+     *
+     * @return A set of CodeList JSON file paths found in the directory.
+     * @throws IOException If an error occurs while reading the resources.
+     * @throws URISyntaxException If the resource path cannot be converted to a URI.
+     */
+    private static Set<String> listCodeListResources() throws IOException, URISyntaxException {
+        URL resourceUrl = ValidateFpMLCodingSchemeImpl.class.getClassLoader().getResource(CODELIST_COLLECTION_PATH);
+        if (resourceUrl == null) {
+            throw new FileNotFoundException("CodeList directory not found: " + CODELIST_COLLECTION_PATH);
+        }
+
+        Path resourcePath = Paths.get(resourceUrl.toURI());
+
+        if (Files.isDirectory(resourcePath)) {
+            // Development mode: List JSON files from the local resources directory
+            try (Stream<Path> stream = Files.list(resourcePath)) {
+                return stream.map(path -> CODELIST_COLLECTION_PATH + "/" + path.getFileName().toString())
+                        .filter(name -> name.endsWith(".json"))
+                        .collect(Collectors.toSet());
+            }
+        } else {
+            // JAR mode: List JSON files inside the JAR
+            String jarPath = resourceUrl.getPath().split("!")[0].replace("file:", "");
+            try (JarFile jarFile = new JarFile(jarPath)) {
+                return jarFile.stream()
+                        .map(JarEntry::getName)
+                        .filter(name -> name.startsWith(CODELIST_COLLECTION_PATH) && name.endsWith(".json"))
+                        .collect(Collectors.toSet());
+            }
+        }
+    }
+
+    /**
+     * Validates whether the provided code exists in the specified domain's CodeList.
+     *
+     * @param code   The code to validate.
+     * @param domain The domain that determines which CodeList should be used.
+     * @return True if the code exists in the corresponding CodeList; otherwise, false.
+     */
+    @Override
+    protected Boolean doEvaluate(String code, String domain) {
+        logger.debug("Triggered doEvaluate with code={} and domain={}", code, domain);
+        CodeList codeList;
+
+        // Check if the domain is already cached
+        if (cache.containsKey(domain)) {
```

Comment:
Suggestion: I would replace the cache with a `ConcurrentHashMap` and use `computeIfAbsent` to make it thread-safe.